File: python/tests/engine/test_iface_routing.py
"""Routing interface files ([FILES] USE INFLOWS / SAVE OUTFLOWS).

End-to-end coverage of the lifecycle wiring added per
``plans/ROUTING_INTERFACE_FILE_INTEGRATION_PLAN_2026-07-01.md``: a model
with ``USE INFLOWS`` receives the file's flows as external inflows, a model
with ``SAVE OUTFLOWS`` writes a legacy-format interface file, and a chained
A→B pair conserves volume at the shared boundary.

All scratch files live in ``python/tests/engine/output/`` so they can be
reviewed after a run (transparent file IO per CLAUDE.md §4.1).
"""

from __future__ import annotations

import os
import unittest

from openswmm.engine import Solver, MassBalance
from openswmm.engine import RoutingTotal

OUTPUT_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), "output")

FLOW_CFS = 2.0
SIM_SECONDS = 3600.0


def _model_inp(receiver: str, outfall: str, files_line: str = "") -> str:
    """Minimal 1-junction / 1-outfall model, optionally with a [FILES] row."""
    return f"""[OPTIONS]
FLOW_UNITS           CFS
FLOW_ROUTING         DYNWAVE
START_DATE           01/01/2026
START_TIME           00:00:00
REPORT_START_DATE    01/01/2026
REPORT_START_TIME    00:00:00
END_DATE             01/01/2026
END_TIME             01:00:00
REPORT_STEP          00:05:00
WET_STEP             00:05:00
DRY_STEP             00:05:00
ROUTING_STEP         0:00:30

[JUNCTIONS]
{receiver}  100.0  10.0  0.0  0.0  0.0

[OUTFALLS]
{outfall}  95.0  FREE

[CONDUITS]
C1  {receiver}  {outfall}  400.0  0.013  0  0

[XSECTIONS]
C1  CIRCULAR  1.5  0  0  0  1

{files_line}
"""


def _iface_file(node: str, flow_cfs: float) -> str:
    """Legacy-format interface file with a constant flow over 3 h."""
    return (
        "SWMM5 Interface File\n"
        "test inflows\n"
        "300  - reporting time step in sec\n"
        "1    - number of constituents as listed below:\n"
        "FLOW CFS\n"
        "1    - number of nodes as listed below:\n"
        f"{node}\n"
        "Node             Year Mon Day Hr  Min Sec FLOW\n"
        f"{node}  2026 01 01 00 00 00 {flow_cfs}\n"
        f"{node}  2026 01 01 03 00 00 {flow_cfs}\n"
    )


def _write(path: str, content: str) -> str:
    with open(path, "w") as f:
        f.write(content)
    return path


def _run(tag: str, inp_text: str) -> Solver:
    inp = _write(os.path.join(OUTPUT_DIR, f"iface_{tag}.inp"), inp_text)
    s = Solver(
        inp,
        os.path.join(OUTPUT_DIR, f"iface_{tag}.rpt"),
        os.path.join(OUTPUT_DIR, f"iface_{tag}.out"),
    )
    s.open()
    s.initialize()
    s.start()
    for _ in s.steps():
        pass
    s.end()
    return s


class TestUseInflows(unittest.TestCase):
    def test_inflow_file_drives_external_inflow(self):
        iface = _write(
            os.path.join(OUTPUT_DIR, "iface_use_in.txt"),
            _iface_file("J1", FLOW_CFS),
        )
        s = _run("use", _model_inp("J1", "O1",
                                   f'[FILES]\nUSE INFLOWS "{iface}"\n'))
        try:
            ext = MassBalance(s).routing_total(RoutingTotal.EXTERNAL)
            expected = FLOW_CFS * SIM_SECONDS  # 7200 ft³
            self.assertAlmostEqual(ext, expected, delta=abs(expected) * 0.10)
        finally:
            s.close()
            s.destroy()

    def test_missing_inflow_file_fails_start(self):
        inp_text = _model_inp(
            "J1", "O1", '[FILES]\nUSE INFLOWS "no_such_iface_file.txt"\n'
        )
        inp = _write(os.path.join(OUTPUT_DIR, "iface_missing.inp"), inp_text)
        s = Solver(
            inp,
            os.path.join(OUTPUT_DIR, "iface_missing.rpt"),
            os.path.join(OUTPUT_DIR, "iface_missing.out"),
        )
        s.open()
        s.initialize()
        try:
            with self.assertRaises(Exception):
                s.start()
        finally:
            try:
                s.close()
            except Exception:
                pass
            s.destroy()


class TestSaveOutflowsAndChaining(unittest.TestCase):
    def test_chained_models_conserve_boundary_volume(self):
        a_in = _write(
            os.path.join(OUTPUT_DIR, "iface_a_in.txt"),
            _iface_file("J1", FLOW_CFS),
        )
        a_out = os.path.join(OUTPUT_DIR, "iface_a_out.txt")

        # Model A: interface inflow at J1 → outfall O1, saved to a_out.
        s_a = _run(
            "chain_a",
            _model_inp(
                "J1", "O1",
                f'[FILES]\nUSE INFLOWS "{a_in}"\nSAVE OUTFLOWS "{a_out}"\n',
            ),
        )
        try:
            a_outflow = MassBalance(s_a).routing_total(RoutingTotal.OUTFLOW)
        finally:
            s_a.close()
            s_a.destroy()
        self.assertGreater(a_outflow, 0.0)

        # The saved file must be a legacy-format interface file.
        with open(a_out) as f:
            content = f.read()
        self.assertTrue(content.startswith("SWMM5 Interface File"))
        self.assertIn("FLOW CFS", content)
        self.assertIn("O1", content)

        # Model B: junction named O1 receives model A's outfall rows.
        s_b = _run(
            "chain_b",
            _model_inp("O1", "OB", f'[FILES]\nUSE INFLOWS "{a_out}"\n'),
        )
        try:
            b_ext = MassBalance(s_b).routing_total(RoutingTotal.EXTERNAL)
        finally:
            s_b.close()
            s_b.destroy()

        self.assertGreater(b_ext, 0.0)
        self.assertAlmostEqual(b_ext, a_outflow, delta=abs(a_outflow) * 0.15)

File: tests/unit/engine/test_iface_routing.cpp
// SPDX-License-Identifier: Apache-2.0
//
// Copyright 2026 Caleb Buahin
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * @file test_iface_routing.cpp
 * @brief End-to-end tests for routing interface files
 *        ([FILES] USE INFLOWS / SAVE OUTFLOWS).
 *
 * @details Verifies the lifecycle wiring added per
 *          plans/ROUTING_INTERFACE_FILE_INTEGRATION_PLAN_2026-07-01.md:
 *          - USE INFLOWS applies the file's flows as node lateral inflows
 *            (booked as external inflow in the routing mass balance).
 *          - SAVE OUTFLOWS produces a legacy-format interface file with a
 *            header and one row per outlet node per reporting step.
 *          - A missing / malformed inflows file fails swmm_engine_start().
 *          - Chained coupling: model A SAVE OUTFLOWS → model B USE INFLOWS
 *            conserves volume at the shared boundary.
 *
 *          Working directory is tests/unit/engine/data/ at runtime (set by
 *          the test CMakeLists.txt); all scratch files are created there
 *          with unique names and removed on success.
 *
 * @see Legacy: src/legacy/engine/iface.c, routing.c addIfaceInflows()
 * @ingroup engine_tests
 *
 * @author   Caleb Buahin <caleb.buahin@gmail.com>
 * @copyright Copyright (c) 2026 Caleb Buahin. All rights reserved.
 * @license  Apache-2.0
 */

#include <gtest/gtest.h>

#include <cstdio>
#include <filesystem>
#include <fstream>
#include <sstream>
#include <string>

#include <openswmm/engine/openswmm_engine.h>
#include <openswmm/engine/openswmm_model.h>
#include <openswmm/engine/openswmm_links.h>
#include <openswmm/engine/openswmm_massbalance.h>
#include <openswmm/engine/openswmm_statistics.h>

namespace fs = std::filesystem;

namespace {

// ---------------------------------------------------------------------------
// Fixture helpers
// ---------------------------------------------------------------------------

// Minimal 1-junction / 1-outfall model. `receiver` names the junction so the
// chained test can receive rows written for another model's outfall.
std::string minimal_model_inp(const std::string& receiver = "J1",
                              const std::string& outfall  = "O1") {
    std::ostringstream s;
    s << "[OPTIONS]\n"
         "FLOW_UNITS           CFS\n"
         "FLOW_ROUTING         DYNWAVE\n"
         "START_DATE           01/01/2026\n"
         "START_TIME           00:00:00\n"
         "REPORT_START_DATE    01/01/2026\n"
         "REPORT_START_TIME    00:00:00\n"
         "END_DATE             01/01/2026\n"
         "END_TIME             01:00:00\n"
         "REPORT_STEP          00:05:00\n"
         "WET_STEP             00:05:00\n"
         "DRY_STEP             00:05:00\n"
         "ROUTING_STEP         0:00:30\n"
         "\n[JUNCTIONS]\n"
         ";;Name  Elev  MaxDepth  InitDepth  SurDepth  Aponded\n"
      << receiver << "  100.0  10.0  0.0  0.0  0.0\n"
      << "\n[OUTFALLS]\n"
         ";;Name  Elev  Type\n"
      << outfall << "  95.0  FREE\n"
      << "\n[CONDUITS]\n"
         ";;Name  From  To  Length  N  InOff  OutOff\n"
         "C1  " << receiver << "  " << outfall << "  400.0  0.013  0  0\n"
      << "\n[XSECTIONS]\n"
         ";;Link  Shape  Geom1  Geom2  Geom3  Geom4  Barrels\n"
         "C1  CIRCULAR  1.5  0  0  0  1\n";
    return s.str();
}

// Legacy-format routing interface file: one node, constant flow, two
// periods bracketing the whole simulation (00:00 → 03:00).
std::string constant_inflow_iface(const std::string& node, double flow_cfs) {
    std::ostringstream s;
    s << "SWMM5 Interface File\n"
         "test inflows\n"
         "300  - reporting time step in sec\n"
         "1    - number of constituents as listed below:\n"
         "FLOW CFS\n"
         "1    - number of nodes as listed below:\n"
      << node << "\n"
      << "Node             Year Mon Day Hr  Min Sec FLOW\n"
      << node << "  2026 01 01 00 00 00 " << flow_cfs << "\n"
      << node << "  2026 01 01 03 00 00 " << flow_cfs << "\n";
    return s.str();
}

void write_file(const std::string& path, const std::string& content) {
    std::ofstream f(path);
    ASSERT_TRUE(f.is_open()) << "cannot create " << path;
    f << content;
}

std::string slurp(const std::string& path) {
    std::ifstream f(path);
    std::stringstream ss;
    ss << f.rdbuf();
    return ss.str();
}

class IfaceRoutingTest : public ::testing::Test {
protected:
    SWMM_Engine engine_ = nullptr;
    std::vector<std::string> scratch_;

    void TearDown() override {
        if (engine_) {
            swmm_engine_close(engine_);
            swmm_engine_destroy(engine_);
            engine_ = nullptr;
        }
        for (const auto& p : scratch_) {
            std::error_code ec;
            fs::remove(p, ec);
        }
    }

    std::string scratch(const std::string& name) {
        auto p = (fs::current_path() / name).string();
        scratch_.push_back(p);
        return p;
    }

    // Open a model from INP text and stop before start().
    void open_model(const std::string& tag, const std::string& inp_text) {
        const auto inp = scratch("iface_" + tag + ".inp");
        const auto rpt = scratch("iface_" + tag + ".rpt");
        const auto out = scratch("iface_" + tag + ".out");
        write_file(inp, inp_text);

        engine_ = swmm_engine_create();
        ASSERT_NE(engine_, nullptr);
        ASSERT_EQ(swmm_engine_open(engine_, inp.c_str(), rpt.c_str(),
                                   out.c_str(), nullptr), SWMM_OK)
            << "open failed: " << swmm_get_last_error_msg(engine_);
        ASSERT_EQ(swmm_engine_initialize(engine_), SWMM_OK);
    }

    void run_to_completion() {
        double elapsed = 0.0;
        for (;;) {
            int rc = swmm_engine_step(engine_, &elapsed);
            ASSERT_EQ(rc, SWMM_OK) << swmm_get_last_error_msg(engine_);
            if (elapsed <= 0.0) break;
        }
        ASSERT_EQ(swmm_engine_end(engine_), SWMM_OK);
    }

    void close_engine() {
        ASSERT_EQ(swmm_engine_close(engine_), SWMM_OK);
        swmm_engine_destroy(engine_);
        engine_ = nullptr;
    }
};

// ---------------------------------------------------------------------------
// USE INFLOWS — flows from the file must arrive as external inflows.
// ---------------------------------------------------------------------------

TEST_F(IfaceRoutingTest, UseInflows_AppliesLateralFlow) {
    const double kFlow = 2.0;   // cfs, constant over the 1 h simulation
    const auto iface = scratch("iface_use.txt");
    write_file(iface, constant_inflow_iface("J1", kFlow));

    open_model("use", minimal_model_inp());
    ASSERT_EQ(swmm_files_set(engine_, "INFLOWS_PATH", iface.c_str()), SWMM_OK);
    ASSERT_EQ(swmm_engine_start(engine_, 1), SWMM_OK)
        << swmm_get_last_error_msg(engine_);
    run_to_completion();

    // 2 cfs × 3600 s = 7200 ft³ booked as external inflow.
    double ext_vol = 0.0;
    ASSERT_EQ(swmm_get_routing_total(engine_, SWMM_ROUTING_EXTERNAL, &ext_vol),
              SWMM_OK);
    EXPECT_NEAR(ext_vol, kFlow * 3600.0, 0.10 * kFlow * 3600.0)
        << "interface inflow volume not delivered";

    // The conduit downstream of the receiving junction must have flowed.
    int c1 = swmm_link_index(engine_, "C1");
    ASSERT_GE(c1, 0);
    double qmax = 0.0;
    ASSERT_EQ(swmm_stat_link_max_flow(engine_, c1, &qmax), SWMM_OK);
    EXPECT_GT(qmax, 0.5 * kFlow);
}

TEST_F(IfaceRoutingTest, UseInflows_ParsedFromFilesSection) {
    // Same scenario, but the path arrives via the [FILES] section instead
    // of the C API — proves the parse → start() wiring end to end.
    const auto iface = scratch("iface_use_sect.txt");
    write_file(iface, constant_inflow_iface("J1", 2.0));

    std::string inp = minimal_model_inp();
    inp += "\n[FILES]\nUSE INFLOWS \"" + iface + "\"\n";
    open_model("use_sect", inp);
    ASSERT_EQ(swmm_engine_start(engine_, 1), SWMM_OK)
        << swmm_get_last_error_msg(engine_);
    run_to_completion();

    double ext_vol = 0.0;
    ASSERT_EQ(swmm_get_routing_total(engine_, SWMM_ROUTING_EXTERNAL, &ext_vol),
              SWMM_OK);
    EXPECT_GT(ext_vol, 0.0);
}

TEST_F(IfaceRoutingTest, UseInflows_MissingFileFailsStart) {
    open_model("missing", minimal_model_inp());
    ASSERT_EQ(swmm_files_set(engine_, "INFLOWS_PATH",
                             "no_such_interface_file.txt"), SWMM_OK);
    EXPECT_NE(swmm_engine_start(engine_, 1), SWMM_OK)
        << "start() must fail when the inflows file cannot be opened";
}

TEST_F(IfaceRoutingTest, UseInflows_BadHeaderFailsStart) {
    const auto iface = scratch("iface_bad.txt");
    write_file(iface, "NOT A SWMM5 Interface File\ngarbage\n");
    open_model("bad", minimal_model_inp());
    ASSERT_EQ(swmm_files_set(engine_, "INFLOWS_PATH", iface.c_str()), SWMM_OK);
    EXPECT_NE(swmm_engine_start(engine_, 1), SWMM_OK)
        << "start() must fail on a malformed interface file header";
}

// ---------------------------------------------------------------------------
// SAVE OUTFLOWS — legacy-format header + one row per outlet per report step.
// ---------------------------------------------------------------------------

TEST_F(IfaceRoutingTest, SaveOutflows_WritesLegacyHeaderAndRows) {
    const auto in_iface  = scratch("iface_chain_in.txt");
    const auto out_iface = scratch("iface_save_out.txt");
    write_file(in_iface, constant_inflow_iface("J1", 2.0));

    open_model("save", minimal_model_inp());
    ASSERT_EQ(swmm_files_set(engine_, "INFLOWS_PATH",  in_iface.c_str()),
              SWMM_OK);
    ASSERT_EQ(swmm_files_set(engine_, "OUTFLOWS_PATH", out_iface.c_str()),
              SWMM_OK);
    ASSERT_EQ(swmm_engine_start(engine_, 1), SWMM_OK)
        << swmm_get_last_error_msg(engine_);
    run_to_completion();
    close_engine();   // flushes + closes the interface file

    ASSERT_TRUE(fs::exists(out_iface));
    const std::string content = slurp(out_iface);

    // Legacy header format.
    EXPECT_EQ(content.rfind("SWMM5 Interface File", 0), 0u)
        << "first line must be the legacy magic string";
    EXPECT_NE(content.find("FLOW CFS"), std::string::npos);
    EXPECT_NE(content.find("number of nodes"), std::string::npos);
    EXPECT_NE(content.find("O1"), std::string::npos)
        << "outfall O1 must be listed as an outlet node";

    // Data rows: 1 h simulation / 5 min reporting → ~12 rows for O1.
    int rows = 0;
    std::istringstream ss(content);
    std::string line;
    bool past_header = false;
    while (std::getline(ss, line)) {
        if (line.rfind("Node", 0) == 0) { past_header = true; continue; }
        if (past_header && line.rfind("O1", 0) == 0) ++rows;
    }
    EXPECT_GE(rows, 10) << "expected ~12 report-step rows for O1";
    EXPECT_LE(rows, 14) << "rows written per routing step instead of per "
                           "reporting step?";
}

TEST_F(IfaceRoutingTest, SameInflowsAndOutflowsFileFailsStart) {
    const auto iface = scratch("iface_same.txt");
    write_file(iface, constant_inflow_iface("J1", 1.0));
    open_model("same", minimal_model_inp());
    ASSERT_EQ(swmm_files_set(engine_, "INFLOWS_PATH",  iface.c_str()), SWMM_OK);
    ASSERT_EQ(swmm_files_set(engine_, "OUTFLOWS_PATH", iface.c_str()), SWMM_OK);
    EXPECT_NE(swmm_engine_start(engine_, 1), SWMM_OK)
        << "legacy ERR 357: inflows and outflows must not share a file";
}

// ---------------------------------------------------------------------------
// Chained coupling: model A SAVE OUTFLOWS → model B USE INFLOWS.
// ---------------------------------------------------------------------------

TEST_F(IfaceRoutingTest, ChainedModels_ConserveBoundaryVolume) {
    const double kFlow = 2.0;
    const auto a_in  = scratch("iface_a_in.txt");
    const auto a_out = scratch("iface_a_out.txt");
    write_file(a_in, constant_inflow_iface("J1", kFlow));

    // --- Model A: J1 → O1, driven by a constant interface inflow. ---
    open_model("chain_a", minimal_model_inp("J1", "O1"));
    ASSERT_EQ(swmm_files_set(engine_, "INFLOWS_PATH",  a_in.c_str()),  SWMM_OK);
    ASSERT_EQ(swmm_files_set(engine_, "OUTFLOWS_PATH", a_out.c_str()), SWMM_OK);
    ASSERT_EQ(swmm_engine_start(engine_, 1), SWMM_OK)
        << swmm_get_last_error_msg(engine_);
    run_to_completion();

    double a_outflow = 0.0;
    ASSERT_EQ(swmm_get_routing_total(engine_, SWMM_ROUTING_OUTFLOW, &a_outflow),
              SWMM_OK);
    ASSERT_GT(a_outflow, 0.0);
    close_engine();

    // --- Model B: junction named O1 receives model A's outfall rows. ---
    open_model("chain_b", minimal_model_inp("O1", "OB"));
    ASSERT_EQ(swmm_files_set(engine_, "INFLOWS_PATH", a_out.c_str()), SWMM_OK);
    ASSERT_EQ(swmm_engine_start(engine_, 1), SWMM_OK)
        << swmm_get_last_error_msg(engine_);
    run_to_completion();

    double b_ext = 0.0;
    ASSERT_EQ(swmm_get_routing_total(engine_, SWMM_ROUTING_EXTERNAL, &b_ext),
              SWMM_OK);
    EXPECT_GT(b_ext, 0.0) << "model B received nothing from model A's file";
    // Boundary conservation within routing tolerance. A's outfall volume
    // ramps up over the first report steps, so allow a generous band.
    EXPECT_NEAR(b_ext, a_outflow, 0.15 * a_outflow);
}

// ---------------------------------------------------------------------------
// Parser parity: wrong-mode rows are fatal input errors (legacy ERR_ITEMS).
// ---------------------------------------------------------------------------

TEST_F(IfaceRoutingTest, SaveInflowsRowFailsOpen) {
    std::string inp = minimal_model_inp();
    inp += "\n[FILES]\nSAVE INFLOWS \"whatever.txt\"\n";

    const auto path = scratch("iface_badmode.inp");
    const auto rpt  = scratch("iface_badmode.rpt");
    const auto out  = scratch("iface_badmode.out");
    write_file(path, inp);

    engine_ = swmm_engine_create();
    ASSERT_NE(engine_, nullptr);
    EXPECT_NE(swmm_engine_open(engine_, path.c_str(), rpt.c_str(),
                               out.c_str(), nullptr), SWMM_OK)
        << "SAVE INFLOWS must be rejected at parse time (legacy ERR_ITEMS)";
}

} // namespace
